ggml-metal.m

```diff
@@ -431,9 +277,9 @@ void ggml_metal_graph_compute(
                 const enum ggml_type src1t = src1 ? src1->type : GGML_TYPE_COUNT;
                 const enum ggml_type dstt  = dst  ? dst->type  : GGML_TYPE_COUNT;
 
-                id<MTLBuffer> id_src0 = src0 ? ggml_metal_get_buffer(ctx, src0, &offs_src0) : nil;
-                id<MTLBuffer> id_src1 = src1 ? ggml_metal_get_buffer(ctx, src1, &offs_src1) : nil;
-                id<MTLBuffer> id_dst  = dst  ? ggml_metal_get_buffer(ctx, dst,  &offs_dst)  : nil;
+                id<MTLBuffer> id_src0 = src0 ? src0->data : nil;
+                id<MTLBuffer> id_src1 = src1 ? src1->data : nil;
+                id<MTLBuffer> id_dst  = dst  ? dst->data  : nil;
```

Comment:
So here, instead of `src0->data` which is the GPU address of the memory, I need to get the Metal buffer that I created in the wrapper. How should I access it?

Reply:
If you cannot use the pointer directly, an option could be to store the offset in `ggml_tensor::data` (just pass a NULL pointer to `ggml_allocator_simple_init`), and store the MTLBuffer in `ggml_tensor::extra`.

Reply:
To set the `extra` pointer, you can implement `init_tensor` in the `ggml_backend_buffer_interface` in the same way as `free_data`. It is called by the allocator after allocating a tensor.

Reply:
If you do this in this way, I think you will also have to handle views differently, because in some cases views are created in a different `ggml_buffer` than their sources. For example, the KV cache has its own `ggml_buffer`, but it is used via views that are created in the compute buffer.
So you would have to consider the OP of a tensor to determine if it is a view, and if so use the `MTLBuffer` of its parent.

It may also be possible to use the `gpuAddress` of `MTLBuffer` in the kernels directly, if you pass the pointer directly instead of going through a `MTLComputeCommandEncoder`. But I may be wrong about that, this is just what I could find in a quick search now.

Reply:
> If you cannot use the pointer directly, an option could be to store the offset in ggml_tensor::data

Hm, this part is not clear - not sure how to get the offset in `init_tensor`

Reply:
The `extra` of the result of `ggml_rope_custom_inplace` seems to remain NULL.
I guess when making views, we have to keep the `extra` as well

Reply:
Right, I forgot about that, inplace ops are effectively views, so that's a problem. In my branch, I have already removed all inplace ops because they also interfere with the improved memory management. In the long term my goal is to make ops inplace automatically in the allocator, and users should use `ggml_cpy` if they actually need the result to be stored in the same tensor.

I think that keeping the `extra` when making views could be a good solution. You can also remove all the inplace ops like I did - there isn't a `ggml_rope_custom` implementation, but it is trivial to add it, I already did in my branch.

Reply:
After removing the inplace ops, now I get a `get_tensor` call where the `tensor->extra` is null.
It is triggered from here:

https://github.com/ggerganov/llama.cpp/blob/d626b5522af84158e94e3b0338d9d4664bdee17f/ggml-backend.c#L468

This is after the ropes in the first layer:

```
dst:  0x15d00f5a0, 114764
ggml_metal_graph_compute_block_invoke: op - ROPE
ggml_metal_graph_compute_block_invoke: src0 -  f32 [  128,    32,     1], 1, tmpq (reshaped)
ggml_metal_graph_compute_block_invoke: dst  -  f32 [  128,    32,     1], 1, Qcur
	copying tensor 0 (Kcur) (16384 bytes)
XXXXXXXXXXXXXXX GET : 103040 0x600002fe83a0, backend = Metal
XXXXXXXXXXXXXXX GET : 0x19280 0x600002fe83a0 0x584004000 offset = 0
	copying tensor 1 (Vcur) (16384 bytes)
XXXXXXXXXXXXXXX GET : 86656 0x0, backend = Metal
GGML_ASSERT: ggml-metal.m:951: tensor->extra != nil && "tensor not allocated"
```

Reply:
Ah, it's likely because `Vcur` is a "view" (i.e. `ggml_transpose`)

Reply:
I agree that dealing with `extra` in `ggml.c` is not good. I will look into a better way to solve this. I think this could possibly be solved by adding a callback similar to `init_tensor` for views, then the backend would have an opportunity to set `extra` there, or any other initialization it may need to do with views.

Reply:
I think I got an idea how to implement this to fit the new interface and keep the Metal buffers in the metal context avoiding the use of `extra`. Will probably take another shot over the weekend

Reply:
If you rebase (I suggest you don't, the allocator needs more work), keep in mind that partial offloading to the GPU is currently broken, only full offloading works.